Buffer block copy analyzer 1

mahdiva opened the pull request:
<!--

Make sure you have read the contribution guidelines: 
- https://docs.microsoft.com/contribute/dotnet/dotnet-contribute-code-analysis#contribute-docs-for-caxxxx-rules
- https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

If your Pull Request is doing one of the following:

- Adding a new diagnostic analyzer or a code fix
- Adding or updating resource strings used by analyzers and code fixes
- Updating analyzer package versions in [Versions.props](../eng/Versions.props)

Then, make sure to run `msbuild /t:pack /v:m` in the repository root; otherwise, the CI build will fail.

Note: Consider merging the PR base branch (`2.9.x`, `main`, or `release/*`) into your branch before you run the pack command to reduce merge conflicts.

-->

ryzngard committed:
Add CSharp analyzer and tests

ryzngard committed:
Add VB analyzer and tests

ryzngard committed:
Fix WaitAll message and title

ryzngard committed:
Add code fix and tests

ryzngard committed:
Add unfixed cases to tests

ryzngard committed:
Run msbuild /t:pack /v:m

ryzngard committed:
Format

ryzngard committed:
Add assert

ryzngard committed:
Apply suggestions from code review

Co-authored-by: Youssef Victor <31348972+Youssef1313@users.noreply.github.com>

ryzngard committed:
Update xlf files

mavasani committed:
Merge pull request #4987 from dotnet/merges/release/6.0.1xx-preview3-to-release/6.0.1xx-preview4

Merge release/6.0.1xx-preview3 to release/6.0.1xx-preview4

mavasani committed:
Merge pull request #4949 from dotnet/merges/release/6.0.1xx-preview1-to-release/6.0.1xx-preview2

Merge release/6.0.1xx-preview1 to release/6.0.1xx-preview2

ryzngard committed:
Remove need for language specific syntax in analyzer

ryzngard committed:
Move to performance instead of usage

ryzngard committed:
Change to CA1839 and CA1840

buyaa-n committed:
Remove unsupported attribute when allow deny list mixed (#5000)

jmarolf committed:
Merge pull request #5004 from dotnet/merges/release/6.0.1xx-preview2-to-release/6.0.1xx-preview3

Merge release/6.0.1xx-preview2 to release/6.0.1xx-preview3

jmarolf committed:
Merge remote-tracking branch 'upstream/main' into merges/main-to-release/6.0.1xx-preview4

jmarolf committed:
Merge pull request #4976 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

jmarolf committed:
Merge pull request #5010 from dotnet/merges/release/6.0.1xx-preview3-to-release/6.0.1xx-preview4

Merge release/6.0.1xx-preview3 to release/6.0.1xx-preview4

jmarolf committed:
Merge pull request #5015 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

buyaa-n committed:
OSX needs to treated  as the same platform as MacOS, everywhere (#5012)

* OSX same as MacOS, need to treated as same everywhere

mavasani committed:
Merge pull request #5017 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

mavasani committed:
Merge branch 'release/6.0.1xx-preview4' into merges/main-to-release/6.0.1xx-preview4

mavasani committed:
Merge pull request #5018 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

jmarolf committed:
Merge pull request #5019 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

ryzngard committed:
Fix IsSingleTaskArgument

mavasani committed:
Merge pull request #5020 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

mavasani committed:
Merge branch 'release/6.0.1xx-preview4' into merges/main-to-release/6.0.1xx-preview4

jmarolf committed:
Merge pull request #5022 from dotnet/merges/main-to-release/6.0.1xx-preview4

Merge main to release/6.0.1xx-preview4

jmarolf committed:
Update Versions.props

jmarolf committed:
Merge pull request #5029 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

Commit:
Change diagnostic id after rebase

Commit:
PR feedback

NewellClark committed:
- Apply requested changes
- Rename CSharpDictionaryContainsMethodsFixer.cs and BasicDictionaryContainsMethodsFixer.vb to be more consistent with other fixer file names

NewellClark committed:
Run msbuild

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into issue-35343

NewellClark committed:
Fix resx tag garbled in merge.

jkoritzinsky committed:
Add analyzer and code fixer for providing a public parameterless SafeHandle constructor.

Outline tests and start filling them in.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/arcade build 20210419.2 (#5039)

[release/6.0.1xx-preview4] Update dependencies from dotnet/arcade

jmarolf committed:
Merge branch 'release/6.0.1xx' into merges/main-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5031 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

NewellClark committed:
Update src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Change doc link to ca1841

NewellClark committed:
Apply requested changes

- Change ID to CA1842

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into issue-33789

NewellClark committed:
Add newline to end of resx file

NewellClark committed:
Add newline character

jkoritzinsky committed:
Fill in the rest of the unit tests.

jkoritzinsky committed:
Undo accidental file rename.

buyaa-n committed:
CA1416 fix ambiguity in Version comparison (#4943)

jmarolf committed:
Merge branch 'release/6.0.1xx' into merges/release/6.0.1xx-preview4-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5042 from dotnet/merges/release/6.0.1xx-preview4-to-release/6.0.1xx

Merge release/6.0.1xx-preview4 to release/6.0.1xx

mavasani committed:
Merge branch 'release/6.0.1xx-preview4' into issue-35343

NewellClark committed:
Fix garbled unshipped.md markup

mavasani committed:
Merge pull request #4687 from NewellClark/issue-35343

(1/4) Issue 35343 - Replace Dictionary<,>.Keys.Contains with Dictionary<,>.ContainsKey

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into issue-33789

jkoritzinsky committed:
PR feedback.

jkoritzinsky committed:
Update src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md

Co-authored-by: Youssef Victor <youssefvictor00@gmail.com>

jkoritzinsky committed:
Coalesce fixer types.

jkoritzinsky committed:
Apply suggestions from code review

Co-authored-by: Youssef Victor <youssefvictor00@gmail.com>

ryzngard committed:
Merge remote-tracking branch 'upstream/release/6.0.1xx-preview4' into feature/task_when_wait_all_analyzer

jkoritzinsky committed:
Use alternative diagnostic location format.

jkoritzinsky committed:
Fix resources based on feedback.

ryzngard committed:
Add rules back to AnalyzerReleases.Unshipped

jkoritzinsky committed:
Remove extra parameter that seems unneeded.

ryzngard committed:
Fix spacing in test

mavasani committed:
Merge pull request #5052 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

ryzngard committed:
Update src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgument.cs

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

ryzngard committed:
Update description wording

ryzngard committed:
Updated documentation based on description

Commit:
Merge pull request #4841 from ryzngard/feature/task_when_wait_all_analyzer

Add WhenAll and WaitAll analyzer for single task argument

Youssef1313 committed:
Cleanup DoNotUseWhenAllOrWaitAllWithSingleArgument

Youssef1313 committed:
Another approach

Commit:
Merge pull request #5060 from Youssef1313/cleanup

Cleanup DoNotUseWhenAllOrWaitAllWithSingleArgument

NewellClark committed:
Change ID to CA1844

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into issue-33789

NewellClark committed:
Fix failing CA

jmarolf committed:
Merge branch 'release/6.0.1xx' into merges/main-to-release/6.0.1xx

jmarolf committed:
Merge remote-tracking branch 'upstream/release/6.0.1xx-preview4' into merges/release/6.0.1xx-preview4-to-release/6.0.1xx

jmarolf committed:
fixup bad merge

NewellClark committed:
Apply formatting changes

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

jmarolf committed:
Merge pull request #5056 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

jmarolf committed:
Merge branch 'release/6.0.1xx' into merges/release/6.0.1xx-preview4-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5048 from dotnet/merges/release/6.0.1xx-preview4-to-release/6.0.1xx

Merge release/6.0.1xx-preview4 to release/6.0.1xx

NewellClark committed:
Apply requested changes

- Apply Mavasani's changes
- Remove unused method from fixers

NewellClark committed:
Change ID to CA1845

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into use-span-based-concat-analyzer

NewellClark committed:
.

NewellClark committed:
Rebuild xlf files to fix CI

jkoritzinsky committed:
Regenerate generated files.

jkoritzinsky committed:
Remove periods.

jkoritzinsky committed:
Update xlf

jkoritzinsky committed:
Add additional test with larger inheritance tree.

jmarolf committed:
Merge pull request #5076 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

jmarolf committed:
Merge pull request #5078 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

jmarolf committed:
Merge pull request #5085 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

NewellClark committed:
Change ID to CA1846

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into prefer-AsSpan-over-Substring-analyzer

NewellClark committed:
Rebuild resource files

NewellClark committed:
Fix CI

NewellClark committed:
.

NewellClark committed:
Force xlf update

NewellClark committed:
Import System namespace correctly

Use case-insensitive string comparison to import System namespace in the VB fixer.

NewellClark committed:
Allow affirmative checks with else-clause

NewellClark committed:
Add VB test

dotnet-bot committed:
Merge main to release/6.0.1xx (#5090)

* Add localization from OneLocBuild (2021-04-27)

* Merge conflicts (#5079)

* Add localization from OneLocBuild (2021-04-28)

Co-authored-by: Jon Fortescue <jonfortescue@protonmail.com>
Co-authored-by: Buyaa Namnan <bunamnan@microsoft.com>
Co-authored-by: Sam Harwell <sam@tunnelvisionlabs.com>
Co-authored-by: Jonathon Marolf <jmarolf@users.noreply.github.com>

buyaa-n committed:
CA1416 Support custom guard members annotated with guard attributes (#5087)

* Support custom guard with attributes

NewellClark committed:
Preserve trivia

- Add tests ensuring trivia is preserved
- Ensure trivia is preserved

NewellClark committed:
Merge branch 'release/6.0.1xx-preview4' into use-CancellationToken-ThrowIfCancellationRequested

NewellClark committed:
Fix CI

NewellClark committed:
Add shared attribute to fixer

NewellClark committed:
Apply suggested changes

Fix possible crash when primitive types can't be loaded

NewellClark committed:
Update src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Update src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Update src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseCancellationTokenThrowIfCancellationRequested.cs

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Simplify fixer

buyaa-n committed:
Merge conflicts

buyaa-n committed:
Fix lint error

jmarolf committed:
Merge pull request #5097 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

NewellClark committed:
Create files and tests

mavasani committed:
Merge pull request #4764 from NewellClark/use-span-based-concat-analyzer

(3/4) Use Span-Based String Concat Analyzer and Fixer

NewellClark committed:
Use simple names in diagnostic messages

- Remove usages of `MinimallyQualifiedFormat`
- Change message so it makes sense with simple names

NewellClark committed:
Merge branch 'release/6.0.1xx' into issue-33789

mavasani committed:
Merge pull request #4726 from NewellClark/issue-33789

(2/4) Override Stream ReadAsync/WriteAsync Analyzer

NewellClark committed:
Update src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Apply suggested changes

NewellClark committed:
Merge branch 'release/6.0.1xx' into prefer-AsSpan-over-Substring-analyzer

NewellClark committed:
Run msbuild to fix CI

NewellClark committed:
Merge branch 'release/6.0.1xx' into use-CancellationToken-ThrowIfCancellationRequested

jkoritzinsky committed:
Merge branch 'release/6.0.1xx' of github.com:dotnet/roslyn-analyzers into safehandle-analyzer-and-fixer

NewellClark committed:
Implement analyzer and fixer

Add GetFirstOrDefaultMemberWithParameterTypes to utilities

NewellClark committed:
Run msbuild

NewellClark committed:
Merge remote-tracking branch 'dotnet/release/6.0.1xx' into #45552-use-string-equals-over-string-compare

NewellClark committed:
Fix build warnings

NewellClark committed:
Fix more build warnings

Remove all references to SyntaxGenerator from analyzer class

NewellClark committed:
Apply suggested changes

NewellClark committed:
Restructure analyzer and fixer

Analyzer and fixer were intertwined in unhealthy ways that also caused CI failures.

NewellClark committed:
Run msbuild

jmarolf committed:
Merge pull request #5113 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

jmarolf committed:
Update Versions.props

move release/6.0.1xx to preview 6

mavasani committed:
Merge pull request #5118 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

NewellClark committed:
Apply suggestions from code review

Co-authored-by: Manish Vasani <mavasani@microsoft.com>

NewellClark committed:
Apply suggested changes

NewellClark committed:
Merge remote-tracking branch 'dotnet/release/6.0.1xx' into #45552-use-string-equals-over-string-compare

mavasani committed:
Merge pull request #4864 from NewellClark/use-CancellationToken-ThrowIfCancellationRequested

Use cancellation token throw if cancellation requested

NewellClark committed:
Merge branch 'release/6.0.1xx' into prefer-AsSpan-over-Substring-analyzer

NewellClark committed:
Change diagnostic id to CA2251

NewellClark committed:
Merge remote-tracking branch 'dotnet/release/6.0.1xx' into #45552-use-string-equals-over-string-compare

mavasani committed:
Merge pull request #5119 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

mavasani committed:
Merge pull request #4806 from NewellClark/prefer-AsSpan-over-Substring-analyzer

(4/4) Prefer 'AsSpan' over 'Substring' analyzer

NewellClark committed:
Merge branch 'release/6.0.1xx' into #45552-use-string-equals-over-string-compare

jkoritzinsky committed:
Don't access the semantic model directly inside RegisterCodeFixesAsync.

jkoritzinsky committed:
Use string.Empty instead of null

NewellClark committed:
Apply suggested changes

NewellClark committed:
Assert if symbols can't be loaded in fixer

Commit:
Merge pull request #5116 from NewellClark/#45552-use-string-equals-over-string-compare

#45552 use string equals over string compare

jkoritzinsky committed:
Merge branch 'release/6.0.1xx' of github.com:dotnet/roslyn-analyzers into safehandle-analyzer-and-fixer

mavasani committed:
Merge pull request #5043 from jkoritzinsky/safehandle-analyzer-and-fixer

Implement Analyzer and best-effort Code Fix for SafeHandle Public Parameterless Constructor guidance.

buyaa-n committed:
[CA1416] Targeted build attributes can be overriden in cross platform assemblies (#5117)

* Targeted build attributes can be overriden in cross platform assemblies

MeikTranel committed:
Implement CA1847 - Use string.Contains(char) whenever possible

Fan out to different implementations for VB and CSharp since named argument have no IOperation abstraction

mavasani committed:
Merge pull request #4908 from MeikTranel/StringContainsCodeFix

Add Analyzer & Codefix - CA1847 -  Use String.Contains(Char) instead of String.Contains(String) with single characters

jmarolf committed:
Merge remote-tracking branch 'upstream/main' into merges/main-to-release/6.0.1xx

jmarolf committed:
Update Versions.props

jmarolf committed:
Update Versions.props

jmarolf committed:
Merge branch 'release/6.0.1xx' into merges/release/6.0.1xx-preview7-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5121 from dotnet/merges/main-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5156 from dotnet/merges/release/6.0.1xx-preview7-to-release/6.0.1xx

Merge release/6.0.1xx-preview7 to release/6.0.1xx

mavasani committed:
Merge pull request #5164 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

jmarolf committed:
Merge remote-tracking branch 'upstream/main' into merges/main-to-release/6.0.1xx

jmarolf committed:
fixup merge conflicts

jmarolf committed:
fixing errors given by the test framework about editorconfig files being additional files

jmarolf committed:
Merge pull request #5167 from dotnet/merges/main-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5196 from dotnet/merges/main-to-release/6.0.1xx

jmarolf committed:
Merge remote-tracking branch 'upstream/main' into merges/main-to-release/6.0.1xx

jmarolf committed:
update baseline for 6.0.x

jmarolf committed:
Merge pull request #5204 from dotnet/merges/main-to-release/6.0.1xx

jmarolf committed:
Merge remote-tracking branch 'upstream/main' into merges/main-to-release/6.0.1xx

jmarolf committed:
Merge pull request #5211 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

dotnet-bot committed:
Merge pull request #5229 from dotnet/merges/release/6.0.1xx-to-release/6.0.1xx-preview7

Merge release/6.0.1xx to release/6.0.1xx-preview7

Commit:
Analyzer to detect RequiresPreviewFeatures attribute (#5155)

* Detect RequiresPreviewFeaturesAttribute in source

* Create new rules

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthAnalyzer.cs:
```diff
@@ -0,0 +1,190 @@
+                        return;
+                    }
+
+                    ImmutableArray<IArgumentOperation> arguments = invocationOperation.Arguments;
```
Can you move this arguments check to before the TargetMethod check? I think the arguments check is cheaper

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthAnalyzer.cs:
```diff
@@ -0,0 +1,190 @@
+                        {
+                            if (srcArg.Value is IConversionOperation srcArgValue)
+                            {
+                                if (lastArg.Instance.GetReferencedMemberOrLocalOrParameter() == srcArgValue.Operand.GetReferencedMemberOrLocalOrParameter())
```
The lines inside this if block are repeated for dstArg. I would move these to another local function and call it from both places

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferConstCharOverConstUnitString.Fixer.cs:
```diff
@@ -69,6 +69,10 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
                             return;
                         }
 
+                        int foo = 1, bar = 1;
```
Delete these lines.

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferConstCharOverConstUnitStringTests.cs:
```diff
@@ -83,7 +83,7 @@ End Module
         public async Task TestMultipleDeclarations()
         {
-            const string multipleDeclarations_cs = @" 
+            const string multipleDeclarations_cs = @"
```
Revert changes to this file

pgovind commented on src/Utilities/Compiler/WellKnownTypeNames.cs:
```diff
@@ -99,7 +99,10 @@ internal static class WellKnownTypeNames
         public const string SystemAttributeUsageAttribute = "System.AttributeUsageAttribute";
         public const string SystemBoolean = "System.Boolean";
         public const string SystemBuffersMemoryManager1 = "System.Buffers.MemoryManager`1";
+        public const string SystemBuffer = "System.Buffer";
```
Nit: These need to be in alphabetical order

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthAnalyzer.cs:
```diff
@@ -0,0 +1,190 @@
+                    IArgumentOperation dstArg = arguments[2];
+                    IArgumentOperation lastArg = arguments[4];
+
+                    bool CheckLengthProperty(IPropertyReferenceOperation lastArg)
```
We should rename this to something more descriptive. Right now it doesn't tell us what the method is checking for. Something like "CheckLengthPropertyOnByteOrSByteArrays" should work.

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthTests.cs:
```diff
@@ -0,0 +1,453 @@
+");
+        }
+
+        [Fact]
```
1 cool thing you can do here is to use a Theory instead of a `Fact`. With Theory you can write these 4 test case (byte, sbyte, src.Length, dst.Length) with just 1 function. See `TestAppendLiteral` in `PreferConstCharOverConstUnitStringTests.cs` to get an idea of how it works and https://andrewlock.net/creating-parameterised-tests-in-xunit-with-inlinedata-classdata-and-memberdata/#:~:text=xUnit%20uses%20the%20%5BFact%5D%20attribute%20to%20denote%20a,the%20most%20common%20is%20with%20an%20%5BInlineData%5D%20attribute. for an overview of `Theory` vs `Fact`

You don't have to make the change right now. I recommend doing this change once you get the PR up in the dotnet/roslyn-analyzers repo and are waiting for reviews.

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthTests.cs:
```diff
@@ -0,0 +1,453 @@
+        }
+
+        [Fact]
+        public async Task NumOfBytesAsMethodInvocation()
```
Add another test case here to test `Buffer.BlockCopy(new int[] {1, 2, 3, 4}, 0, new int[] {5, 6, 7, 8}, 0, someLength)` ?

pgovind reviewed (commented):
Analyzer itself looks good. Just a few minor comments. Once those are fixed we can put it up in the roslyn-analyzers repo

mahdiva committed:
Add Buffer.BlockCopy Length Analyzer

pgovind commented on src/Utilities/Compiler/WellKnownTypeNames.cs:
```diff
@@ -98,8 +98,11 @@ internal static class WellKnownTypeNames
         public const string SystemBoolean = "System.Boolean";
+        public const string SystemBuffer = "System.Buffer";
         public const string SystemBuffersMemoryManager1 = "System.Buffers.MemoryManager`1";
+        public const string SystemArray = "System.Array";
```
Did you forget to push changes to this file? `SystemArray` needs to go further to the top I think? And `SystemSByte` would go down to where the rest of the `SystemS*` things are

mahdiva committed:
Fix alphabetical order

pgovind commented:
Yup this looks good to me now. I think you can open a PR against dotnet/roslyn-analyzers now

maryamariyan committed:
Add Logging Analyzer

maryamariyan committed:
Convert to IOperation analysis

maryamariyan committed:
Updated tests based on refactor

mahdiva committed:
Address PR comments

RikkiGibson committed:
Fix concurrency issue in CA1069 analyzer (#5243)

maryamariyan committed:
minor code cleanup

maryamariyan committed:
PR feedback

dotnet-bot committed:
Merge pull request #5249 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

maryamariyan committed:
Use CreateDiagnostic extension method

dotnet-bot committed:
Merge pull request #5236 from dotnet/merges/release/6.0.1xx-to-release/6.0.1xx-preview7

Merge release/6.0.1xx to release/6.0.1xx-preview7

mahdiva committed:
Change category to Reliability

dotnet-bot committed:
Merge pull request #5250 from dotnet/merges/release/6.0.1xx-to-release/6.0.1xx-preview7

Merge release/6.0.1xx to release/6.0.1xx-preview7

xtqqczze committed:
Delete IDE dispose analyzer rules

The IDE dispose analyzer rules have been deleted from Roslyn:

https://github.com/dotnet/roslyn/commit/eeba499ecf839ec35bca25062d69d2fc5c4885b9

maryamariyan committed:
update auto generated files

maryamariyan committed:
Fix "Run markdownlint" error

maryamariyan committed:
isPortedFxCopRule set false by default

RikkiGibson committed:
Remove dead code in CA1069 analyzer (#5253)

maryamariyan committed:
isReportedAtCompilationEnd set to false

mavasani committed:
Merge pull request #5255 from xtqqczze/patch-1

Delete IDE dispose analyzer rules from .editorconfig

bbbriana committed:
Initial add of richnav indexing

mahdiva committed:
Merge branch 'release/6.0.1xx' into BufferBlockCopyAnalyzer_1

bbbriana committed:
Update yml spacing

bbbriana committed:
Remove unauthorized settings

bbbriana committed:
Update pool

mahdiva committed:
Refactor

sharwell committed:
Update MicrosoftCodeAnalysisVersionForTests to 4.0.0-2.final

sharwell committed:
Add Net60 reference assemblies for tests

sharwell committed:
Update TestPreviewLanguageFeatures for new functionality

Evangelink committed:
Remove some empty statements

bbbriana committed:
Attempt to fix test step in build pipeline

bbbriana committed:
Add richnav nuget source

bbbriana committed:
RichNav run worked in main yml -- removing from there and copying it to richnav solo yml

bbbriana committed:
Remove added whitespace

bbbriana committed:
Remove unused vars, rename displayName for clarity

sharwell committed:
Rename test class to match file name

sharwell committed:
Update INameOfOperation handling to use correct constant value

sharwell committed:
Update IOperation to understand params arguments

sharwell committed:
Merge pull request #5278 from Evangelink/empty-statements

Remove some empty statements

sharwell committed:
Merge pull request #5274 from sharwell/update-test-dependency

Update test dependency

jmarolf committed:
Merge pull request #5267 from brianaeng/breng/addRichNav

dotnet-bot committed:
Merge pull request #5258 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

dotnet-bot committed:
Merge pull request #5285 from dotnet/merges/release/6.0.1xx-to-release/6.0.1xx-preview7

Merge release/6.0.1xx to release/6.0.1xx-preview7

jmarolf committed:
Update Versions.props

maryamariyan committed:
move analyzer from csharp to core

maryamariyan committed:
Apply API review feedback.

- Switch category to Reliability for fifth diagnostic
- Switch severity to hidden from disabled for first two

maryamariyan committed:
narrow down scope of 4th diagnostic to LogTrace and friends

Commit:
Fix TODOs and update unit tests with diagnostics (#5287)

* Fix TODOs and update unit tests with diagnostics

* Address comments

* sq

* Address feedback

maryamariyan committed:
updating messages to become a bit more clear

maryamariyan committed:
Apply nit feedback

maryamariyan committed:
nit feedback

maryamariyan committed:
if to else if

dotnet-bot committed:
Merge pull request #5294 from dotnet/merges/main-to-release/6.0.1xx

Merge main to release/6.0.1xx

maryamariyan committed:
rename: named hole -> named placeholder

maryamariyan committed:
fix paranthesis issue

jmarolf committed:
Merge remote-tracking branch 'upstream/release/6.0.1xx-preview7' into merges/release/6.0.1xx-preview7-to-release/6.0.1xx

jmarolf committed:
do not update the label

jmarolf committed:
Merge pull request #5244 from maryamariyan/logging-az

dotnet-bot committed:
Merge pull request #5293 from dotnet/merges/release/6.0.1xx-preview7-to-release/6.0.1xx

Merge release/6.0.1xx-preview7 to release/6.0.1xx

mahdiva committed:
Address PR comments

mahdiva committed:
New comments

buyaa-n committed:
CA1416 MacCatalyst is a superset of iOS (#5266)

* MacCatalyst is superset of Ios

jeffhandley committed:
Implement the ModuleInitializerAttributeShouldNotBeUsedInLibraries analyzer (#5347)

mahdiva committed:
New unit tests

mahdiva committed:
Merge remote-tracking branch 'upstream/release/6.0.1xx' into BufferBlockCopyAnalyzer_1

mahdiva committed:
NIT

mahdiva committed:
Fix build errors

mahdiva committed:
Skip Array.CreateInstance unit test
